### DISCUSSION TITLE
Build Bazel

### DISCUSSION THREAD

**Discussion by Mame29** [termux/termux-packages#14607] [General]
i am getting an error while i am building the bezel. this is the error display
`ld.lld: error: undefined symbol: __android_log_write`

![Screenshot_20230117_043713](https://user-images.githubusercontent.com/72637907/212770169-931be95b-e7f5-4f05-80f8-7d7ce25ce860.jpg)

How to fix it ?

**Comment by hnmn** [comment]
U need to set `-llog` flag ig, i also try to build bazel long time ago its build system is complex maybe open package request issue.
but best of luck with bazel.

**Reply by Mame29** [reply]
it's work thanks, but i found errors
![Screenshot_20230117_115525](https://user-images.githubusercontent.com/72637907/212821441-1813674c-0e86-43cd-8b26-8a94aef60eb9.jpg)
![Screenshot_20230117_125723](https://user-images.githubusercontent.com/72637907/212821476-875f47a5-ae8d-4f8f-940c-22262d92df85.jpg)

Now how fixed ?

**Reply by hnmn** [reply]
Bazel have its own lib sources which needs patch to works on android,also u need to add -llog flag in 
cmake file. like i said i didn't understand its build system u need deep dive into how bazel build from source. 
make package request possible dev will add package to termux.

**Reply by Mame29** [reply]
requests package already available at #7450

**Reply by Mame29** [reply]
until now not finished

**Reply by hnmn** [reply]
https://github.com/bazelbuild/bazel/blob/414824173363e579d34afc1aa16bc97a220743dc/scripts/bootstrap/compile.sh#L57

Try to add `-llog` and `-landroid-spawn` above line 

I also don't no much just look into this file how files are build and search bezel on arm on net 
U get some references

**Reply by Mame29** [reply]
i didn't edit the file, but added some arguments in `EXTRA_BAZEL_ARGS` and `BAZEL_JAVAC_OPTS`

**Reply by Mame29** [reply]
i got a new error
`/data/data/com.termux/files/home/bazel/bazel-4.2.2/src/BUILD:319:9: Executing genrule //src:embedded_tools_nojdk failed: (Exit 127): bash failed: error executing command`
![Screenshot_20230117_145217](https://user-images.githubusercontent.com/72637907/212840569-836c79c8-f1c1-4bf1-a180-5181829d07cd.jpg)

i don't know whare is error ,,
do you have any suggestions ?

**Reply by Mame29** [reply]
i try use termux-chroot

**Reply by hnmn** [reply]
> i got a new error `/data/data/com.termux/files/home/bazel/bazel-4.2.2/src/BUILD:319:9: Executing genrule //src:embedded_tools_nojdk failed: (Exit 127): bash failed: error executing command` ![Screenshot_20230117_145217](https://user-images.githubusercontent.com/72637907/212840569-836c79c8-f1c1-4bf1-a180-5181829d07cd.jpg)
> 
> i don't know whare is error ,, do you have any suggestions ?

Yeah try termux-chroot
Bazel add default env like /bin/bash 
That's why u got error

**Reply by hnmn** [reply]
> i try use termux-chroot

Are you able build successfully now?

**Reply by Mame29** [reply]
No, i got now error again
`src/main/tools/process-tools.cc:220:13: error: use of undeclared identifier 'wait3'`
![Screenshot_20230117_200416](https://user-images.githubusercontent.com/72637907/212906408-4d8ca56c-5ba1-43c8-bb91-c65025e858fa.jpg)
how i included in to cflags ?

**Reply by Mame29** [reply]
> saya tidak mengedit file, tetapi menambahkan beberapa argumen di `EXTRA_BAZEL_ARGS`dan`BAZEL_JAVAC_OPTS`

sorry I forgot, I added the code in the script/boostrap/compile.sh file


```patch
--- anu/scripts/bootstrap/compile.sh    1980-01-01 00:00:00.000000000 +0700
+++ scripts/bootstrap/compile.sh        2023-01-18 00:32:13.862792792 +0700
@@ -436,6 +436,7 @@                                                              
    done                                                                                                                                                        

    "${JAVA_HOME}/bin/java" \                                                  
+      --add-opens=java.base/java.lang=ALL-UNNAMED \                                 
       -XX:+HeapDumpOnOutOfMemoryError -Xverify:none -Dfile.encoding=ISO-8859-1 \                                                                                  
       -XX:HeapDumpPath=${OUTPUT_DIR} \                                              
       -Djava.util.logging.config.file=${OUTPUT_DIR}/javalog.properties \
```

**Comment by Mame29** [comment]
command compiler bazel version 4.2.2
`termux-chroot 'env EXTRA_BAZEL_ARGS="--host_javabase=@local_jdk//:jdk --jobs=8 --host_linkopt="-llog" --host_linkopt="-landroid-spawn" --host_linkopt="-landroid-shmem" --host_copt="-D_XOPEN_SOURCE" --host_copt="-fpic" --action_env=PREFIX --action_env=CFLAGS --action_env=LDFLAGS --copt="-D_XOPEN_SOURCE" --copt="-I/data/data/com.termux/files/usr/include" --linkopt="-llog" --linkopt="-landroid-shmem" --linkopt="-landroid-spawn" --linkopt="-L/data/data/com.termux/files/usr/aarch64-linux-android/lib" --linkopt="-lc"" ./compile.sh'`

**Reply by hnmn** [reply]
Is this command work for you

**Reply by Mame29** [reply]
Yes, but there are some edited files

**Reply by hnmn** [reply]
Are you able run bazel, if yes can you share binary

**Reply by Mame29** [reply]
yes is run 
![Screenshot_20230118_221651](https://user-images.githubusercontent.com/72637907/213210787-59edbb9f-75b3-4228-973d-020da3538d81.jpg)
![Screenshot_20230118_221824](https://user-images.githubusercontent.com/72637907/213210881-114a92b3-0440-4d2c-af6c-5c7130655c08.jpg)

**Reply by hnmn** [reply]
Great 👍
Can please you share patches which files you edited or upload your bazel binary to your GitHub i like to test it.

**Reply by Mame29** [reply]
I have made a package request at #14654

**Reply by Mame29** [reply]
@hnmn 
I have build bazel for aarch64 and i add patch file please check https://github.com/Mame29/termux-bazel

**Reply by hnmn** [reply]
great, running fine ran some examples.

**Comment by Mame29** [comment]
patch

```patch
--- src/main/tools/process-tools.cc.old 2023-01-18 15:10:35.054531683 +0700
+++ src/main/tools/process-tools.cc     2023-01-18 15:00:00.314531645 +0700
@@ -217,7 +217,11 @@
     // Discard any zombies that we may get when the child subreaper feature is
     // enabled.
     do {
+#if defined(HAVE_WAIT3)
       err = wait3(&status, 0, rusage);
+#else
+      err = wait4(-1, &status, 0, rusage);
+#endif
     } while (err != pid || (err == -1 && errno == EINTR));
   } else {
     do

--- src/main/native/unix_jni_linux.cc.old       2023-01-18 15:55:26.121269042 +0700
+++ src/main/native/unix_jni_linux.cc   2023-01-18 17:41:32.233319245 +0700
@@ -31,7 +31,7 @@
   // functionality which is not compatible with not the
   // SUSv3-conformant one which returns an error code; see DESCRIPTION
 // at strerror(3).
-#if !__GLIBC__ || (_POSIX_C_SOURCE >= 200112L && !_GNU_SOURCE)
+#if !__GLIBC__ && !__USE_GNU || (_POSIX_C_SOURCE >= 200112L && !_GNU_SOURCE)
   if (strerror_r(error_number, buf, sizeof buf) == -1) {
     return std::string("");
   } else {

--- src/main/tools/BUILD.old    2023-01-18 20:18:31.106736472 +0700
+++ src/main/tools/BUILD        2023-01-18 20:20:18.618736479 +0700
@@ -76,12 +76,7 @@
         "//src/conditions:openbsd": ["dummy-sandbox.c"],
         "//src/conditions:windows": ["dummy-sandbox.c"],
         "//conditions:default": [
-            "linux-sandbox.cc",
-            "linux-sandbox.h",
-            "linux-sandbox-options.cc",
-            "linux-sandbox-options.h",
-            "linux-sandbox-pid1.cc",
-            "linux-sandbox-pid1.h",
+            "dummy-sandbox.c",
         ],
     }),
     linkopts = select({
@@ -99,10 +94,7 @@
         "//src/conditions:freebsd": [],
         "//src/conditions:openbsd": [],
         "//src/conditions:windows": [],
-        "//conditions:default": [
-            ":logging",
-            ":process-tools",
-        ],
+        "//conditions:default": [],
     }),
 )

--- ./src/java_tools/buildjar/java/com/google/devtools/build/buildjar/VanillaJavaBuilder.java.old       2023-01-18 21:14:47.506736674 +0700
+++ ./src/java_tools/buildjar/java/com/google/devtools/build/buildjar/VanillaJavaBuilder.java   2023-01-18 21:17:52.394736685 +0700
@@ -78,7 +78,7 @@
   private FileSystem getJarFileSystem(Path sourceJar) throws IOException {
     FileSystem fs = filesystems.get(sourceJar);
     if (fs == null) {
-      filesystems.put(sourceJar, fs = FileSystems.newFileSystem(sourceJar, null));
+      filesystems.put(sourceJar, fs = FileSystems.newFileSystem(sourceJar, (ClassLoader) null));
     }
     return fs;
   }

--- anu/scripts/bootstrap/compile.sh    1980-01-01 00:00:00.000000000 +0700
+++ scripts/bootstrap/compile.sh        2023-01-18 00:32:13.862792792 +0700
@@ -436,6 +436,7 @@                                                              
    done                                                                                                                                                        

    "${JAVA_HOME}/bin/java" \                                                  
+      --add-opens=java.base/java.lang=ALL-UNNAMED \                                 
       -XX:+HeapDumpOnOutOfMemoryError -Xverify:none -Dfile.encoding=ISO-8859-1 \                                                                                  
       -XX:HeapDumpPath=${OUTPUT_DIR} \                                              
       -Djava.util.logging.config.file=${OUTPUT_DIR}/javalog.properties \
```
